File: app/src/main/java/org/mifos/mobilebanking/api/services/LoanAccountsListService.java
package org.mifos.mobilebanking.api.services;

import org.mifos.mobilebanking.api.ApiEndPoints;
import org.mifos.mobilebanking.models.accounts.loan.LoanAccount;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithdraw;
import org.mifos.mobilebanking.models.payload.LoansPayload;
import org.mifos.mobilebanking.models.templates.loans.LoanTemplate;
import org.mifos.mobilebanking.utils.Constants;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;
import retrofit2.http.Query;

/**
 * @author Vishwajeet
 * @since 23/6/16.
 */
public interface LoanAccountsListService {

    @GET(ApiEndPoints.LOANS + "/{loanId}/")
    Observable<LoanAccount> getLoanAccountsDetail(@Path("loanId") long loanId);

    @GET(ApiEndPoints.LOANS + "/{loanId}")
    Observable<LoanWithAssociations> getLoanWithAssociations(
            @Path("loanId") long loanId,
            @Query("associations") String associationType);

    @GET(ApiEndPoints.LOANS + "/template?templateType=individual")
    Observable<LoanTemplate> getLoanTemplate(@Query("clientId") long clientId);

    @GET(ApiEndPoints.LOANS + "/template?templateType=individual")
    Observable<LoanTemplate> getLoanTemplateByProduct(
            @Query("clientId") long clientId,
            @Query("productId") Integer productId);

    @POST(ApiEndPoints.LOANS)
    Observable<ResponseBody> createLoansAccount(@Body LoansPayload loansPayload);

    @PUT(ApiEndPoints.LOANS + "/{loanId}/")
    Observable<ResponseBody> updateLoanAccount(@Path("loanId") long loanId,
                                              @Body LoansPayload loansPayload);

    @POST(ApiEndPoints.LOANS + "/{loanId}?command=withdrawnByApplicant")
    Observable<ResponseBody> withdrawLoanAccount(@Path(Constants.LOAN_ID) long loanId,
                                                 @Body LoanWithdraw loanWithdraw);
}

File: core/network/src/commonMain/kotlin/org/mifos/mobile/core/network/services/LoanAccountsListService.kt
/*
 * Copyright 2026 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/mobile-mobile/blob/master/LICENSE.md
 */
package org.mifos.mobile.core.network.services

import de.jensklingenberg.ktorfit.http.Body
import de.jensklingenberg.ktorfit.http.GET
import de.jensklingenberg.ktorfit.http.POST
import de.jensklingenberg.ktorfit.http.PUT
import de.jensklingenberg.ktorfit.http.Path
import de.jensklingenberg.ktorfit.http.Query
import io.ktor.client.statement.HttpResponse
import kotlinx.coroutines.flow.Flow
import org.mifos.mobile.core.network.dto.loanAccount.LoanAccountResponseDto
import org.mifos.mobile.core.network.dto.loanAccount.LoanWithAssociationsResponseDto
import org.mifos.mobile.core.network.dto.payloads.LoanAccountApplicationPayloadDto
import org.mifos.mobile.core.network.dto.payloads.LoanWithdrawPayloadDto
import org.mifos.mobile.core.network.dto.templates.loan.LoanTemplateResponseDto
import org.mifos.mobile.core.network.dto.transaction.LoanTransactionDetailsResponseDto
import org.mifos.mobile.core.network.utils.ApiEndPoints

interface LoanAccountsListService {
    @GET(ApiEndPoints.LOANS + "/{loanId}/")
    fun getLoanAccountsDetail(@Path("loanId") loanId: Long): Flow<LoanAccountResponseDto>?

    @GET(ApiEndPoints.LOANS + "/{loanId}")
    fun getLoanWithAssociations(
        @Path("loanId") loanId: Long,
        @Query("associations") associationType: String?,
    ): Flow<LoanWithAssociationsResponseDto>

    @GET(ApiEndPoints.LOANS + "/template?templateType=individual")
    fun getLoanTemplate(@Query("clientId") clientId: Long?): Flow<LoanTemplateResponseDto>

    @GET(ApiEndPoints.LOANS + "/template?templateType=individual")
    fun getLoanTemplateByProduct(
        @Query("clientId") clientId: Long?,
        @Query("productId") productId: Int?,
    ): Flow<LoanTemplateResponseDto>

    @POST(ApiEndPoints.LOANS)
    suspend fun createLoansAccount(@Body loansPayload: LoanAccountApplicationPayloadDto?): HttpResponse

    @PUT(ApiEndPoints.LOANS + "/{loanId}/")
    suspend fun updateLoanAccount(
        @Path("loanId") loanId: Long,
        @Body loansPayload: LoanAccountApplicationPayloadDto?,
    ): HttpResponse

    @POST(ApiEndPoints.LOANS + "/{loanId}?command=withdrawnByApplicant")
    suspend fun withdrawLoanAccount(
        @Path("loanId") loanId: Long,
        @Body loanWithdraw: LoanWithdrawPayloadDto?,
    ): HttpResponse

    @GET(ApiEndPoints.LOANS + "/{loanId}/transactions/{transactionId}")
    fun getLoanTransactionDetails(
        @Path("loanId") loanId: Long,
        @Path("transactionId") transactionId: Long,
    ): Flow<LoanTransactionDetailsResponseDto>
}
